src/optimizer.cpp

```diff
@@ -124,67 +101,75 @@ geometry_msgs::msg::TwistStamped Optimizer::evalControl(
   const geometry_msgs::msg::Twist & robot_speed,
   const nav_msgs::msg::Path & plan, nav2_core::GoalChecker * goal_checker)
 {
-  for (size_t i = 0; i < settings_.iteration_count_; ++i) {
+  state_.pose = robot_pose;
+  state_.speed = robot_speed;
+
+  bool stop_flag = false;
+  costs_.fill(0);
+
+  for (size_t i = 0; i < settings_.iteration_count; ++i) {
     generated_trajectories_ =
-      generateNoisedTrajectories(robot_pose, robot_speed);
-    auto && costs = critic_manager_.evalTrajectoriesScores(
-      state_, generated_trajectories_, plan, robot_pose, goal_checker);
-    updateControlSequence(costs);
+      generateNoisedTrajectories();
+
+
+    auto data =
+      models::CriticFunctionData{state_, generated_trajectories_,
+      utils::toTensor(plan), goal_checker, costs_, stop_flag, std::nullopt};
+
+    critic_manager_.evalTrajectoriesScores(data);
+    updateControlSequence();
+  }
+
+  auto control = getControlFromSequenceAsTwist(plan.header.stamp);
+
+  if (settings_.shift_control_sequence) {
+    shiftControlSequence();
   }
 
-  auto control = getControlFromSequenceAsTwist(
-    settings_.control_sequence_shift_offset_,
-    plan.header.stamp);
 
-  shiftControlSequence();
+  if (stop_flag) {
+    reset();
```

Comment:
If we fail, shouldn't we notify the navigation task that it failed to generate trajectory to be able to react with recoveries?

Reply:
I think we'd also for this method want to `reset()`, but its important to let the navigation task know if it could not generate a valid trajectory to execute recoveries. I think you should `throw std::runtime_error(...)` here which the controller server will pick up on so you don't return a bogus `control`

Reply:
add throw

Reply:
I don't see it added

Reply:
i remove it then, after some conclusions in another thread. We need here double check for failure

Reply:
added fallback behavior